### PR TITLE
Allow for R recovery

### LINKED ISSUE #1585: Reveal R mechanism
At the moment, R's are only revealed for ongoing sessions.

This, however, is fragile as R's are only revealed when the consumer is connected to us. If we want a more resilient mechanism, a job should start when providing a service and periodically check for promises that might need to be revealed and revealing them.

The session pingpong would then also need to be redesigned a little to make the synchronous calls to accountant obsolete.

### PR DESCRIPTION
**Pull request by vkuznecovas**
The provider will now send an encrypted R to the accountant so that he can store it.

Once there, the encrypted R will be sent back to us if a missmatch occurs.

Upon receiving such an error, a provider will decrypt the R and send it back to accountant.

closes #1585

### CODE REVIEW

**Comment by vkuznecovas** [comment]
The required changes are already live on accountant and are backwards compatible.

File: session/pingpong/accountant_caller.go [hunk 1]
```diff
@@ -206,4 +211,27 @@ var accountantCauseToError = map[string]error{
...
+}
+
+var encryptedRRegexp = regexp.MustCompile(`Encrypted recovery data: "(?P<recoverydetails>[0-9a-f]+)"`)
+
+type rRecoveryDetails struct {
+	R           string `json:"r"`
+	AgreementID uint64 `json:"agreement_id"`
+}
+
+func parseErrNeedsRRecoveryDetails(in string) (string, error) {
```
Comment (Waldz) [inline session/pingpong/accountant_caller.go]: Looks bogus. Why parsing from error message is needed at all?

Comment (vkuznecovas) [reply session/pingpong/accountant_caller.go]: Because I don't want to do extra requests.

Comment (tadaskay) [reply session/pingpong/accountant_caller.go]: I think what Waldz meant to ask is: perhaps it's possible to use a more convenient struct than a string?

Comment (anjmao) [reply session/pingpong/accountant_caller.go]: Agree. Maybe returning recovery details in separate field could help?

File: session/pingpong/accountant_caller.go [hunk 2]
```diff
@@ -29,13 +30,65 @@ import (
...
+func (aer *AccountantErrorResponse) UnmarshalJSON(data []byte) error {
+	var s struct {
+		CausedBy     string `json:"cause"`
+		ErrorMessage string `json:"message"`
+		ErrorData    string `json:"data"`
+	}
+
+	err := json.Unmarshal(data, &s)
+	if err != nil {
+		return err
```
Comment (zolia) [inline session/pingpong/accountant_caller.go]: would wrap another context here.

File: session/pingpong/invoice_tracker.go [hunk 3]
```diff
@@ -271,9 +279,25 @@ func (it *InvoiceTracker) requestPromise(r []byte, pm crypto.ExchangeMessage) er
...
 	}
 
+	details := rRecoveryDetails{
+		R:           hex.EncodeToString(r),
+		AgreementID: pm.AgreementID,
+	}
+
+	bytes, err := json.Marshal(details)
+	if err != nil {
+		return errors.Wrap(err, "could not marshal R recovery details")
```
Comment (zolia) [inline session/pingpong/invoice_tracker.go]: old wrap?

File: session/pingpong/invoice_tracker.go [hunk 4]
```diff
@@ -486,30 +510,64 @@ func (it *InvoiceTracker) waitForInvoicePayment(hlock []byte) {
...
-		ErrAccountantNotFound,
-		ErrAccountantNoPreviousPromise,
-		ErrAccountantMalformedJSON:
+		stdErr.Is(err, ErrAccountantInternal),
+		stdErr.Is(err, ErrAccountantNotFound),
+		stdErr.Is(err, ErrAccountantNoPreviousPromise),
+		stdErr.Is(err, ErrAccountantMalformedJSON):
 		// these are ignorable, we'll eventually fail
 		if it.incrementAccountantFailureCount() > it.deps.MaxAccountantFailureCount {
 			return err
```
Comment (zolia) [inline session/pingpong/invoice_tracker.go]: We should wrap context here.

File: session/pingpong/invoice_tracker.go (hunk 3 above)
Comment (vkuznecovas) [reply session/pingpong/invoice_tracker.go]: This whole component needs to be changed, will do it in a separate pr someday.

File: session/pingpong/invoice_tracker.go (hunk 4 above)
Comment (vkuznecovas) [reply session/pingpong/invoice_tracker.go]: the context is added above.

**Pull request merged**